include/multipass/terminal.h

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (C) 2019 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef MULTIPASS_TERMINAL_H
+#define MULTIPASS_TERMINAL_H
+
+#include <istream>
+#include <memory>
+#include <ostream>
+
+namespace multipass
+{
+class Terminal
+{
+public:
+    virtual ~Terminal();
+
+    virtual std::istream& cin() = 0;
+    virtual std::ostream& cout() = 0;
+    virtual std::ostream& cerr() = 0;
+
+    virtual int cin_fd() const = 0;
+    virtual bool cin_is_tty() const = 0;
+
+    virtual int cout_fd() const = 0;
+    virtual bool cout_is_tty() const = 0;
+
+    using UPtr = std::unique_ptr<Terminal>;
+    static UPtr make_terminal();
+};
+} // namespace multipass
```

Comment:
Naming here confused me a bit: "how can a terminal's cout not be a tty?" I don't have a better name for the class (shell is inappropriate, console is taken, virtual terminal would only confuse things). But perhaps the `cin_is_tty` could be `is_interactive` and `cout_is_tty` could be `is_live`? I don't know, just a thought, maybe it would be even worse. WDYT?

Reply:
"how can a terminal's cout not be a tty?" - when you do this: multipass list | cat, multipass's cout is not a tty, but cin is. Similarly, echo "" | multipass, cin is not a tty, but cout is.

Naming, I'm not sure. My first thought seeing "is_interactive" is that it implies both cin and cout are backed by a tty. But I didn't want this to be too clever. 
"cin/cout_is_live" maybe?

Reply:
OK, my confusion was probably caused more by intrinsic complexity than how is was presented. Yeah go with what you think is best.